File: ahammer3-cs50-2018-x-vigenere-20181004T104216Z/ahammer3-cs50-2018-x-vigenere-20181004T104216Z/vigenere.c
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* This program shifts a string of letters by a key word from plaintext to ciphertext */

int main(int argc, string argv[])
{
    //check if the # arguments is 2
    if (argc != 2)
    {
        printf("Try again - incorrect command line parameters \n");
        return 1;
    }

    //get key from command line argument, check for isalpha, initiate counter and kshifter afterwards
    //counter starts at -1 to shift to 0 at start
    string k = argv[1];
    int lenK = strlen(k);

    for (int i = 0; i < lenK; i++)
    {
        if (!isalpha(k[i]))
        {
            printf("Try again - incorrect command line parameters \n");
            return 1;
        }
    }
    int counter = -1;
    int kshifter;

    //prompt user for plaintext
    printf("plaintext: ");
    string plaintext = get_string();

    //for each character in plaintext
    //if alphabetic, preserve case, shift by key and print ciphertext
    //else we print the character and do not shift by the key, and move on
    printf("ciphertext: ");
    for (int i = 0, n = strlen(plaintext); i < n; i++)
    {
        if (isalpha(plaintext[i]))
        {
            counter = (counter + 1) % lenK;
            if (isupper(k[counter]))
            {
                kshifter = 65;
            }
            else
            {
                kshifter = 97;
            }

            if (isupper(plaintext[i]))
            {
                int shifter = 65;
                printf("%c", (((plaintext[i] + (k[counter] - kshifter)) - shifter) % 26) + shifter);
            }
            else if (islower(plaintext[i]))
            {
                int shifter = 97;
                printf("%c", (((plaintext[i] + (k[counter] - kshifter)) - shifter) % 26) + shifter);
            }
        }
        else
        {
            printf("%c", plaintext[i]);
        }
    }
    printf("\n");

    return 0;
}

File: ahammer3-cs50-2018-x-sentimental-vigenere-20181107T192741Z/ahammer3-cs50-2018-x-sentimental-vigenere-20181107T192741Z/vigenere.py
# This program shifts a string of letters by a key word from plaintext to ciphertext

# import modules
import cs50
import sys

# Check if valid - a function!


def valid(k):
    for char in k:
        if not char.isalpha():
            return False
    # Removed else statement here - it would return true at the first alpha character and ruin rest of process
    return True


def main():

    # Use valid and len to check arguments
    if len(sys.argv) != 2 or not valid(sys.argv[1]):
        print("Try again - incorrect command line parameters")
        sys.exit(1)

    # Set up k and plaintext, remember variable counter
    k = sys.argv[1]
    plaintext = cs50.get_string("plaintext: ")
    counter = 0

    # Set up ciphertext
    print("ciphertext: ", end="")

    # print actual ciphertext based on parameters
    for char in plaintext:
        if not char.isalpha():
            print(char, end="")
            continue  # like in caesar for non alpha characters

        if char.isupper():
            shifter = 65
        else:
            shifter = 97

        plain = ord(char) - shifter
        kshifter = ord(k[counter % len(k)].upper()) - 65
        cipher = (plain + kshifter) % 26
        counter += 1

        print(chr(cipher + shifter), end="")

    print()

    return 0


if __name__ == "__main__":
    main()

# NOTE: In the future, any programs involving functions like valid would probably be better with a def main() in the program and
# calling the following too: if __name__ == "__main__":\n    main()

# NOTE 2: updated this.